notebooks/work-in-progress/dagster_prototypes/multi_assets_etl.py

```diff
@@ -0,0 +1,224 @@
+    op,
+    repository,
+    with_resources,
+)
+
+
+class SQLiteIOManager(IOManager):
+    """IO manager that stores and retrieves values from a sqlite database..
+
+    Args:
+        output_dir (Optional[str]): base directory where all the step outputs which use this object
+            manager will be stored in.
+    """
+
+    def __init__(self, output_dir: Path = None):
+        self.output_dir = output_dir
+
+    def handle_output(self, context, obj):
+        """Write dataframe to a sqlite database."""
+        if context.has_asset_key:
+            # TODO: raise a more helpful exception or use a different dagster function to get the table name.
+            table_name = context.get_asset_identifier()[0]
+        else:
+            table_name = context.get_identifier()
+
+        filepath = self.output_dir / "pudl.sqlite"
+        filepath.touch()
+
+        engine = sa.create_engine(f"sqlite:///{filepath}")
+        print(table_name, obj)
+        with engine.connect() as con:
+            # TODO: Delete data from existing schema and append the new data
+            obj.to_sql(table_name, con, if_exists="replace", index=False)
+
+    def load_input(self, context):
+        # upstream_output.name is the name given to the Out that we're loading for
+        if context.has_asset_key:
+            # TODO: raise a more helpful exception or use a different dagster function to get the table name.
+            table_name = context.get_asset_identifier()[0]
+        else:
+            table_name = context.get_identifier()
+
+        filepath = self.output_dir / "pudl.sqlite"
+
+        engine = sa.create_engine(f"sqlite:///{filepath}")
+        with engine.connect() as con:
+            return pd.read_sql_table(table_name, con)
+
+
+@io_manager(
+    config_schema={
+        "output_dir": Field(
+            str, default_value=str(pathlib.Path().absolute() / "output_dir")
+        )
+    }
+)
+def sqlite_io_manager(init_context):
+    output_dir = Path(init_context.resource_config["output_dir"])
+    output_dir.mkdir(exist_ok=True)
+    return SQLiteIOManager(output_dir)
+
+
+#############
+## Toy FERC extraction
+#############
+@op(
+    out={
+        "raw_steam": Out(pd.DataFrame, io_manager_key="sqlite_io_manager"),
+        "raw_fuel": Out(pd.DataFrame, io_manager_key="sqlite_io_manager"),
+    }
+)
+def dbf2sqlite():
+    return pd.DataFrame([1, 2, 3], columns=["data"]), pd.DataFrame(
+        [1, 2, 3], columns=["data"]
+    )
+
+
+@graph(out={"raw_steam": GraphOut(), "raw_fuel": GraphOut()})
+def etl():
+    """Run pudl_etl job."""
+    return dbf2sqlite()
+
+
+#############
+## Toy PUDL ETL
+#############
+@op(out=Out(io_manager_key="fs_io_manager"))
+def _etl_ferc1(raw_steam) -> dict[str, pd.DataFrame]:
+    return {"steam": pd.DataFrame([1, 2, 3], columns=["data"])}
+
+
+@op(out=Out(io_manager_key="fs_io_manager"))
+def _etl_eia() -> dict[str, pd.DataFrame]:
+    return {"gens": pd.DataFrame([1, 2, 3], columns=["data"])}
```

Comment:
Why do these use filesystem IO managers while the others use SQLite IO managers?

Reply:
These use the filesystem IO manager because they are dictionaries of data frames that we don't want to persist in the database.